## Training on AWS-EMR

This is the actual training of the whole sparkify dataset on a 5 node cluster on AWS-EMR with a Pyspark kernel.

`1.` __Installing and Importing Packages__

`1.1` Installing Packages

In [2]:
#checking spark version

spark.version

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

'2.4.7-amzn-1'

In [1]:
# installing Packages
sc.install_pypi_package("pandas") 
sc.install_pypi_package("matplotlib", "https://pypi.org/simple") #I

VBox()

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

  Using cached https://files.pythonhosted.org/packages/51/51/48f3fc47c4e2144da2806dfb6629c4dd1fa3d5a143f9652b141e979a8ca9/pandas-1.2.4-cp37-cp37m-manylinux1_x86_64.whl
  Using cached https://files.pythonhosted.org/packages/d4/70/d60450c3dd48ef87586924207ae8907090de0b306af2bce5d134d78615cb/python_dateutil-2.8.1-py2.py3-none-any.whl

  Using cached https://files.pythonhosted.org/packages/ce/63/74c0b6184b6b169b121bb72458818ee60a7d7c436d7b1907bd5874188c55/matplotlib-3.4.1-cp37-cp37m-manylinux1_x86_64.whl
  Using cached https://files.pythonhosted.org/packages/8a/bb/488841f56197b13700afd5658fc279a2025a39e22449b7cf29864669b15d/pyparsing-2.4.7-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/33/34/542152297dcc6c47a9dcb0685eac6d652d878ed3cea83bf2b23cb988e857/Pillow-8.2.0-cp37-cp37m-manylinux1_x86_64.whl
  Using cached https://files.pythonhosted.org/packages/f7/d2/e07d3ebb2bd7af696440ce7e754c59dd546ffe1bbe732c8ab68b9c834e61/cycler-0.10.0-py2.py3-none-any.whl
  Using ca

In [2]:
# importing libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from datetime import datetime

import re

from pyspark.sql import SparkSession
from pyspark.sql.types import StringType, DateType, IntegerType
from pyspark.sql.functions import concat, lit, avg, split, isnan, when, count, col, sum, mean, stddev, min, max, round, udf, to_date, datediff 
from pyspark.sql import Window

from pyspark.ml.feature import StringIndexer, VectorAssembler, Normalizer, StandardScaler, MinMaxScaler, OneHotEncoder, StringIndexer
from pyspark.ml.classification import LogisticRegression, GBTClassifier, NaiveBayes, RandomForestClassifier
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
from pyspark.mllib.evaluation import BinaryClassificationMetrics, MulticlassMetrics
from pyspark.mllib.util import MLUtils

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

`2.` __Setting up a spark session__

In [ ]:
# configuring the session

%%configure -f 

{ "conf":{
          "spark.pyspark.python": "python3", "spark.pyspark.virtualenv.enabled": "true", "spark.pyspark.virtualenv.type": "native", "spark.pyspark.virtualenv.bin.path": "/usr/bin/virtualenv", "driverMemory": "6000M"
         }
}

In [9]:
sc.list_packages()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Package                    Version  
-------------------------- ---------
beautifulsoup4             4.9.3    
boto                       2.49.0   
click                      7.1.2    
cycler                     0.10.0   
jmespath                   0.10.0   
joblib                     1.0.1    
kiwisolver                 1.3.1    
lxml                       4.6.2    
matplotlib                 3.4.1    
mysqlclient                1.4.2    
nltk                       3.5      
nose                       1.3.4    
numpy                      1.16.5   
pandas                     1.2.4    
Pillow                     8.2.0    
pip                        9.0.1    
py-dateutil                2.2      
pyparsing                  2.4.7    
python-dateutil            2.8.1    
python37-sagemaker-pyspark 1.4.1    
pytz                       2021.1   
PyYAML                     5.4.1    
regex                      2021.3.17
setuptools                 28.8.0   
six                        1.13.0   
t

In [34]:
# number of rows in the DataFrame
df = spark.read.json('s3n://udacity-dsnd/sparkify/mini_sparkify_event_data.json')
df.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

26259199

`3` __Feature Engineering__

`3.1` Extracting useful features

In [3]:
def clean_data(df):
    
    """
    This functions removes all the rows where userId is empty or Null and 
    returns the dataframe.
    
    Input : DataFrame
    Output : cleaned DataFrame
    """
    
    df_new = df.filter(df["userId"] != "")
    df_new = df.filter(col('userId').isNull()==False)

    
    return df_new

def prepare_dataset(df): 
    
    """
    This function will prepare the DataFrame for Machine Learning
    by Extracting out the useful features, and engineering more 
    relevant features. The unique users will be kept in one of the DataFrames
    which will be used for ML
    
    Input : DataFrame
    Output : DataFrame to be used for applying Machine Learning and the modified input DataFrame
    """
    
    
    
    df = clean_data(df)
    
    
    """Defining churn"""
    cancellation_event = udf(lambda x: 1 if x == "Cancellation Confirmation" else 0, IntegerType())   
    df = df.withColumn("churn", cancellation_event("page"))   
    cancelled_users = df.select(['userId']).where(df.churn == 1).groupby('userId').count().toPandas()['userId'].values
    
    #Filling all the users who pressed the 'Cancellation Confirmation' button with 1
    def fill_array(userId, features):
        if(userId in cancelled_users): return 1
        else : return 0
        
    fill_array_udf = udf(fill_array, IntegerType())
    df = df.withColumn("churn", fill_array_udf(col("userId"), col("churn")))
    
    
    
    
    w = Window.partitionBy('userId') #Partitioning the Data by User Id
    df = df.withColumn('last_ts', max('ts').over(w)) #create last timestamp
    df = df.withColumn('first_ts', min('ts').over(w)) #create first timestamp
    
    #This function will convert timestamp to date
    def get_date_from_ts(ts):
        return str(datetime.utcfromtimestamp(ts / 1000).strftime('%Y-%m-%d'))
    
    get_date_from_ts_udf = udf(get_date_from_ts, StringType())
    df = df.withColumn('last_date', get_date_from_ts_udf(col('last_ts'))) #converting last timestamp to date
    df = df.withColumn('first_date', get_date_from_ts_udf(col('first_ts'))) #converting first timestamp to date
    
    
    df = df.withColumn('date', get_date_from_ts_udf(col('ts'))) #converting all timestamps to date
    
    df = df.withColumn('last_level',when(df.last_ts == df.ts, df.level)) #defining the last level (paid or free) of a user
    
    # create column avg_songs to calculate average number of songs per day
    # first grouping on unique (userId, date) pair and then taking average 
    # over all the dates for a particular user
    w = Window.partitionBy('userId', 'date')
    songs = df.where(df.page == 'NextSong').select('userId', 'date', count('userId').over(w).alias('songs')).distinct()
    w = Window.partitionBy('userId')
    songs = songs.withColumn('avg_songs', avg('songs').over(w))
    songs = songs.select(col("userId").alias("songs_userId"), 'avg_songs')
    songs = songs.withColumn("avg_songs", round(songs["avg_songs"], 2))
    
    # create column avg_events to calculate average number of events per day
    # first grouping on unique (userId, date) pair and then taking average 
    # over all the dates for a particular user
    w = Window.partitionBy('userId', 'date')
    events = df.select('userId', 'date', count('userId').over(w).alias('events')).distinct()
    w = Window.partitionBy('userId')
    events = events.withColumn('avg_events', avg('events').over(w))
    events = events.select(col("userId").alias("events_userId"), 'avg_events')
    events = events.withColumn("avg_events", round(events["avg_events"], 2))
    
    # calculate number of thumbs up for a user
    w = Window.partitionBy('userId')
    thumbsup = df.where(df.page == 'Thumbs Up').select('userId', count('userId').over(w).alias('thumbs_up')).distinct()
    thumbsup = thumbsup.select(col("userId").alias("thumbsup_userId"), 'thumbs_up')
    
    # calculate number of thumbs down for a user
    w = Window.partitionBy('userId')
    thumbsdown = df.where(df.page == 'Thumbs Down').select('userId', count('userId').over(w).alias('thumbs_down')).distinct()
    thumbsdown = thumbsdown.select(col("userId").alias("thumbsdown_userId"), 'thumbs_down')
    
    # calculate days since the date of the first event
    df = df.withColumn("days_active", 
              datediff(to_date(lit(datetime.now().strftime("%Y-%m-%d %H:%M"))),
                       to_date("first_date","yyyy-MM-dd")))
    
    # add column with state of the event based on location column
    def get_state(location):
        location = location.split(',')[-1].strip()
        if (len(location) > 2):
            location = location.split('-')[-1].strip()
    
        return location
    
    get_state_udf = udf(get_state, StringType())
    df = df.withColumn('state', get_state_udf(col('location')))
    
    #add column with last location of the user
    df = df.withColumn('last_state',when(df.last_ts == df.ts, df.state))
    
    # calculate number of add friends for a user
    w = Window.partitionBy('userId')
    addfriend = df.where(df.page == 'Add Friend').select('userId', count('userId').over(w).alias('addfriend')).distinct()
    addfriend = addfriend.select(col("userId").alias("addfriend_userId"), 'addfriend')

    # assemble everything into resulting dataset
    df_ml = df.select('userId', 'gender', 'churn', 'last_level', 'days_active', 'last_state')\
    .dropna().drop_duplicates()
    df_ml = df_ml.join(songs, df_ml.userId == songs.songs_userId).distinct()
    df_ml = df_ml.join(events, df_ml.userId == events.events_userId).distinct()
    df_ml = df_ml.join(thumbsup, df_ml.userId == thumbsup.thumbsup_userId, how='left').distinct()
    df_ml = df_ml.fillna(0, subset=['thumbs_up'])
    df_ml = df_ml.join(thumbsdown, df_ml.userId == thumbsdown.thumbsdown_userId, how='left').distinct()
    df_ml = df_ml.fillna(0, subset=['thumbs_down'])
    df_ml = df_ml.join(addfriend, df_ml.userId == addfriend.addfriend_userId, how='left').distinct()
    df_ml = df_ml.fillna(0, subset=['addfriend'])
    df_ml = df_ml.drop('songs_userId','events_userId', 'thumbsup_userId', 'thumbsdown_userId', 'addfriend_userId')
    
    return df, df_ml
    
df = spark.read.json('s3n://udacity-dsnd/sparkify/mini_sparkify_event_data.json')
df.persist()

df, df_ml = prepare_dataset(df)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [70]:
#Total number of unique users in the Big Dataset

df_ml.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

22263

In [6]:
# number of customers who churn 
df_ml.select('churn').where(col('churn')==1).count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

5002

Therefore the number of customers who churn are 5002 and ones who did not are (22263-5002)=17261
`class imbalance of 3.45:1`. Therefore F1  score will be used as a metric of accuracy

In [31]:
#Total number of columns in the DataFrame
df.columns

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

['artist', 'auth', 'firstName', 'gender', 'itemInSession', 'lastName', 'length', 'level', 'location', 'method', 'page', 'registration', 'sessionId', 'song', 'status', 'ts', 'userAgent', 'userId', 'churn', 'last_ts', 'first_ts', 'last_date', 'first_date', 'date', 'last_level', 'days_active', 'state', 'last_state']

In [60]:
#total number of rows in the DataFrame after cleaning

df.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

25475717

In [20]:
#checking spark memory for this configuration
sc._conf.get('spark.driver.memory')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

'2048M'

`3.2` Indexing of String columns and One Hot Encoding

In [28]:
# index and encode categorical features gender, level

stringIndexerGender = StringIndexer(inputCol="gender", outputCol="genderIndex", handleInvalid = 'skip')
stringIndexerLevel = StringIndexer(inputCol="last_level", outputCol="levelIndex", handleInvalid = 'skip')


# OneHotEncoding these features, using OneHotEncoderEstimator because it has the transform method (updated from OneHotEncoder)
encoder_gender = OneHotEncoderEstimator(inputCols=["genderIndex"], outputCol=["genderVec"])
encoder_level = OneHotEncoderEstimator(inputCols=["levelIndex"], outputCol=["levelVec"])


# create vector for features
features = ['genderVec', 'levelVec', 'days_active', 'avg_songs', 'avg_events', 'thumbs_up', 'thumbs_down', 'addfriend']
assembler = VectorAssembler(inputCols=features, outputCol="features")

# initialize random forest classifier
rf = RandomForestClassifier(labelCol="label", featuresCol="features", numTrees=10, maxDepth=5)

# assemble pipeline
pipeline = Pipeline(stages = [stringIndexerGender, stringIndexerLevel, encoder_gender,encoder_level, assembler, rf])

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

`4` __Applying Machine Learning__

`4.1` splitting the data and fitting the model

In [4]:
df_ml = df_ml.withColumnRenamed("churn", "label")

train, test_valid = df_ml.randomSplit([0.6, 0.4], seed = 42)
test, validation = test_valid.randomSplit([0.5, 0.5], seed = 42)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [29]:
model = pipeline.fit(train)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

`4.2` Observing Model Performance 

In [30]:
pred_train = model.transform(train)
pred_test = model.transform(test)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [31]:
predictionAndLabels = pred_train.rdd.map(lambda lp: (float(lp.prediction), float(lp.label)))

# Instantiate metrics object
metrics = MulticlassMetrics(predictionAndLabels)

# F1 score
print("F1 score on train dataset is %s" % metrics.fMeasure())

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

F1 score on train dataset is 0.793822913550703

In [32]:
predictionAndLabels = pred_test.rdd.map(lambda lp: (float(lp.prediction), float(lp.label)))

# Instantiate metrics object
metrics = MulticlassMetrics(predictionAndLabels)

# F1 score
print("F1 score on test dataset is %s" % metrics.fMeasure())

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

F1 score on test dataset is 0.7878650667874123

We see thant the F1 score on training Dataset is `0.79` and on Test Dataset is `0.78` for the `Random Forest Model`

In [77]:
model_rf = model

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [33]:
#saving the random forest model
model.save('s3://aws-emr-resources-816555935147-us-east-2/notebooks/model-rf')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

`4.3` Setting up a pipeline for Gradient Boosted Trees model 

In [5]:
# index and encode categorical features gender, level and state

stringIndexerGender = StringIndexer(inputCol="gender", outputCol="genderIndex", handleInvalid = 'skip')
stringIndexerLevel = StringIndexer(inputCol="last_level", outputCol="levelIndex", handleInvalid = 'skip')

encoder_gender = OneHotEncoder(inputCol="genderIndex", outputCol="genderVec")
encoder_level = OneHotEncoder(inputCol="levelIndex", outputCol="levelVec")

features = ['genderVec', 'levelVec', 'days_active', 'avg_songs', 'avg_events', 'thumbs_up', 'thumbs_down', 'addfriend']
assembler = VectorAssembler(inputCols=features, outputCol="features")

gbt = GBTClassifier(labelCol="label", featuresCol="features", maxIter=10, maxDepth=5)

pipeline = Pipeline(stages = [stringIndexerGender, stringIndexerLevel, encoder_gender,encoder_level, assembler, gbt])

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [6]:
#fitting the gbt model
model = pipeline.fit(train)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

`4.4` Testing Model Performance

In [7]:
pred_train = model.transform(train)
pred_test = model.transform(test)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [8]:
predictionAndLabels = pred_train.rdd.map(lambda lp: (float(lp.prediction), float(lp.label)))

# Instantiate metrics object
metrics = MulticlassMetrics(predictionAndLabels)

# F1 score
print("F1 score on train dataset is %s" % metrics.fMeasure())

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Exception in thread cell_monitor-6:
Traceback (most recent call last):
  File "/mnt/notebook-env/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/mnt/notebook-env/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/mnt/notebook-env/lib/python3.7/site-packages/awseditorssparkmonitoringwidget-1.0-py3.7.egg/awseditorssparkmonitoringwidget/cellmonitor.py", line 178, in cell_monitor
    job_binned_stages[job_id][stage_id] = all_stages[stage_id]
KeyError: 610



F1 score on train dataset is 0.8093030212384086

In [9]:
predictionAndLabels = pred_test.rdd.map(lambda lp: (float(lp.prediction), float(lp.label)))

# Instantiate metrics object
metrics = MulticlassMetrics(predictionAndLabels)

# F1 score
print("F1 score on test dataset is %s" % metrics.fMeasure())

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

F1 score on test dataset is 0.8066561014263075

We see thant the F1 score on training Dataset is `0.809` and on Test Dataset is `0.806` for the `Random Forest Model`. It is performing slightly better that Random Forest model so I am using it for deployment purposes.

In [10]:
model.save('s3://aws-emr-resources-816555935147-us-east-2/notebooks/model-gbt')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

# Conclusion

In this notebook we went through the process of Engineering the features of the __large dataset containing 26 million rows and then extracting the unique users.The resulting DataSet that we get is about 22 thousand rows which is suitable for applying Machine Learning__

I have applied two models -Random Forest and Gradient Boosted trees and find that the performance of GBT is slightly better. I have saved both the trained models into the S3 bucket and then synced the bucket with a local folder.
__I will be using the GBT model for the case of application based on Flask. Thereafter I will be deploying the model on a web-server like AWS.__

